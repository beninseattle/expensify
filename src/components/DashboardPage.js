import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';

const DashboardPage = () => (
  <div>
    <ExpenseListFilter/>
    <ExpenseList/>
  </div>
);

export default DashboardPage;