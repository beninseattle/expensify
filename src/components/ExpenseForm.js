import React, {Component} from 'react';
import moment from 'moment';
// react dates must be initialized before importing
import 'react-dates/initialize';
import {SingleDatePicker} from 'react-dates';
import Expense from '../models/expense';

/**
 * @class ExpenseForm
 * @property {Object} props
 * @property {onSubmitExpenseCallback} props.onSubmit
 * @property {Expense} props.expense
 * @property {Object} state
 * @property {Expense} state.expense
 */
export default class ExpenseForm extends Component {
  onDescriptionChange = (e) => {
    let newExpense = new Expense(this.state.expense);
    newExpense.description = e.target.value;
    this.setState(() => ({expense: newExpense}));
  };
  onNoteChange = (e) => {
    let newExpense = new Expense(this.state.expense);
    newExpense.note = e.target.value;
    this.setState(() => ({expense: newExpense}));
  };
  // TODO: Fix this input
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      let newExpense = new Expense(this.state.expense);
      newExpense.amount = parseFloat(amount) * 100;
      this.setState(() => ({expense: newExpense}));
    }
  };

  /**
   * @param {moment} createdAt
   */
  onDateChange = (createdAt) => {
    if (createdAt) {
      let newExpense = new Expense(this.state.expense);
      newExpense.createdAt = createdAt.valueOf();
      this.setState(() => ({expense: newExpense}));
    }
  };
  onFocusChange = ({focused}) => {
    this.setState(() => ({calendarFocused: focused}));
  };
  handleSubmit = (e) => {
    e.preventDefault();

    if (!this.state.expense.isValid()) {
      this.setState(() => ({error: 'A description and amount is required'}));
    } else {
      this.setState(() => ({error: ''}));
      this.props.onSubmit(this.state.expense);
    }
  };

  /**
   * @param {Object} props
   * @property {onSubmitExpenseCallback} props.onSubmit
   * @property {Expense} [props.expense]
   */
  constructor(props) {
    super(props);
    this.submitButton = 'Save changes';
    let expense = props.expense;
    if (expense === undefined) {
      expense = new Expense();
      this.submitButton = 'Save expense';
    }
    this.state = {
      expense: expense,
      calendarFocused: false,
      error: ''
    };
  };

  render() {
    return (
      <form className="form" onSubmit={this.handleSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <input
          className="text-input"
          type="text"
          placeholder="Description"
          name="description"
          autoFocus
          value={this.state.expense.description}
          onChange={this.onDescriptionChange}
        />
        <input
          className="text-input"
          type="text"
          placeholder="Amount"
          name="amount"
          value={this.state.expense.amountAsFloat()}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={moment(this.state.expense.createdAt)}
          onDateChange={this.onDateChange}
          focused={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          className="textarea"
          name="note"
          placeholder="Add a note (optional)"
          value={this.state.expense.note}
          onChange={this.onNoteChange}/><br/>
        <div>
          <button className="button">{this.submitButton}</button>
        </div>
      </form>
    );

  }
}