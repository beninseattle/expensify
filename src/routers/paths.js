const paths = {
  login: '/',
  dashboard: '/dashboard',
  addExpense: '/create',
  editExpense: '/edit/',
  help: '/help'
};

export default paths;