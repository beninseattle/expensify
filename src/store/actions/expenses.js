import database from '../../firebase/firebase';
import Expense from '../../models/expense';

/**
 * @callback expenseAction
 * @param {Expense} expense
 */
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

/**
 * Returns function to start async action to add a new expense in Firebase and then dispatch to redux
 * @param {Expense} expense
 * @returns {function(func,func)}
 */
export const startAddExpense = (expense) => {
  // function functionality thanks to redux-thunk
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).push(expense.dataForSave())
      .then((ref) => {
        expense.saveToStore(ref.key);
        dispatch(addExpense(expense));
      })
      .catch((error) => {
        console.log(`Add expense for expense ${expense.id} failed: ${error}`);
      });
  };
};

/**
 * @callback expenseAction
 * @param {Expense} expense - the expense to remove from the store
 */
export const deleteExpense = (expense) => ({
  type: 'DELETE_EXPENSE',
  expense
});

/**
 * Returns function to start async action to delete a specific expense in Firebase and then dispatch to redux
 * @param expense
 * @returns {function(func, func)}
 */
export const startDeleteExpense = (expense) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${expense.id}`).remove()
      .then(() => {
        dispatch(deleteExpense(expense));
      })
      .catch((error) => {
        console.log(`Delete for expense ${expense.id} failed: ${error}`);
      });
  };
};

/**
 * @callback expenseAction
 * @param {Expense} expense - object with 1 or more of the fields of an expense
 */
export const editExpense = (expense) => ({
  type: 'EDIT_EXPENSE',
  expense
});

/**
 * Returns function to start async action to save a specific expense in Firebase and then dispatch to redux
 * @param expense
 * @returns {function(func, func)}
 */
export const startEditExpense = (expense) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${expense.id}`).set(expense.dataForSave())
      .then(() => {
        dispatch(editExpense(expense));
      })
      .catch((error) => {
        console.log(`Edit for expense ${expense.id} failed: ${error}`);
      });
  };
};

/**
 * @callback setExpenses
 */
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

/**
 * Returns function to start async action to load expenses from Firebase and then dispatch to redux
 * @returns {function(func, func)}
 */
export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).once('value')
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((childSnapshot) => {
          expenses.push(new Expense({
            id: childSnapshot.key,
            ...childSnapshot.val()
          }));
        });

        dispatch(setExpenses(expenses));
      })
      .catch((error) => {
        console.log(`Set expenses failed to load data: ${error}`);
      });
  };
};