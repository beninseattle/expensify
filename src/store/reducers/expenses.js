const expensesReducerDefaultState = [];
export const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense];

    case 'DELETE_EXPENSE':
      return state.filter(({id}) => id !== action.expense.id);

    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.equals(action.expense.id)) {
          return action.expense;
        } else {
          return expense;
        }
      });

    case 'SET_EXPENSES':
      return action.expenses;

    default:
      return state;
  }
};
