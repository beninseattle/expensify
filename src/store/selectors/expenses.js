import moment from 'moment';

/**
 * Filtering by dates compares based on a granularity of days
 *
 * @param expenses
 * @param filters
 */
export default (expenses, filters) => {
  let visibleExpenses = expenses.filter((expense) => {
    if (filters.text.length > 0) {
      if (expense.note.toLowerCase().indexOf(filters.text.toLowerCase()) === -1 &&
        expense.description.toLowerCase().indexOf(filters.text.toLowerCase()) === -1) {
        return false;
      }
    }
    if (filters.startDate !== undefined) {
      if (moment(expense.createdAt).isBefore(filters.startDate, 'day')) {
        return false;
      }
    }
    if (filters.stopDate !== undefined) {
      if (moment(expense.createdAt).isAfter(filters.stopDate, 'day')) {
        return false;
      }
    }
    return true;
  });

  visibleExpenses.sort((a, b) => {
    if (filters.sortBy === 'date') {
      if (a.createdAt === b.createdAt) {
        return 0;
      } else {
        return a.createdAt < b.createdAt ? -1 : 1;
      }
    } else if (filters.sortBy === 'amount') {
      if (a.amount === b.amount) {
        return 0;
      } else {
        return a.amount < b.amount ? -1 : 1;
      }
    }
    return 0;
  });
  return visibleExpenses;
};
