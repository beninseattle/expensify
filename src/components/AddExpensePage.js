import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../store/actions/expenses';
import paths from '../routers/paths';

/**
 * @property {Object} props
 * @property {mapDispatchToProps} props.onSubmit
 * @property {Object} props.history
 * @property {function(string)} props.history.push
 */
export class AddExpensePage extends React.Component {
  /**
   * @callback onSubmitExpenseCallback
   * @param expense
   */
  onSubmit = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push(paths.dashboard);
  };

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit}/>
      </div>
    );
  }
}

/**
 * @callback
 * @param {function(expenseAction)} dispatch
 * @returns {{addExpense: (function(Expense): *)}}
 */
const mapDispatchToProps = (dispatch) => {
  return {
    addExpense: (expense) => dispatch(addExpense(expense))
  };
};
export default connect(undefined, mapDispatchToProps)(AddExpensePage);