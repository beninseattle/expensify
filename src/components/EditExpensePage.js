import React from 'react';
import {connect} from 'react-redux';
import {Redirect} from 'react-router-dom';
import ExpenseForm from './ExpenseForm';
import {editExpense, startDeleteExpense} from "../store/actions/expenses";
import paths from '../routers/paths';

/**
 * @property {Object} props
 * @property {Expense} props.expense
 * @property {function(Expense)} props.editExpense
 * @property {function(Expense)} props.deleteExpense
 * @property {Object} props.history
 * @property {function(string)} props.history.push
 * @property {function(string)} props.history.redirect
 */
export class EditExpensePage extends React.Component {
  /**
   * @callback onSubmitExpenseCallback
   * @param {Expense} expense
   */
  onSubmit = (expense) => {
    this.props.editExpense(expense);
    this.props.history.push(paths.dashboard);
  };
  /**
   * @callback onDeleteExpenseCallback
   */
  onDelete = () => {
    this.props.deleteExpense(this.props.expense);
    this.props.history.push(paths.dashboard);
  };

  render() {
    if( this.props.expense === undefined ) {
      return (
        <Redirect to={paths.dashboard} />
      );
    }

    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button onClick={this.onDelete}>delete
        </button>
      </div>
    );
  }
}
/**
 *
 * @param {Object} state
 * @property {Expense[]} state.expenses
 * @param {Object} props
 * @property {string} props.match.params.id
 * @returns {{expense: T}}
 */
const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.equals(props.match.params.id))
});
/**
 * @callback
 * @param {function(expenseAction)} dispatch
 * @returns {{addExpense: (function(Expense)), deleteExpense: (function(Expense))}}
 */

const mapDispatchToProps = (dispatch) => ({
  editExpense: (expense) => dispatch(editExpense(expense)),
  deleteExpense: (expense) => dispatch(startDeleteExpense(expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);