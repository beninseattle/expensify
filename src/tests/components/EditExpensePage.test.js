import React from 'react';
import {shallow} from 'enzyme';
import {EditExpensePage} from "../../components/EditExpensePage";
import expenses from '../fixtures/expenses';
import paths from '../../routers/paths';

let expense, editExpense, deleteExpense, history, wrapper;

beforeEach(() => {
  expense = expenses[0];
  editExpense = jest.fn();
  deleteExpense = jest.fn();
  history = {push: jest.fn()};
});

test('should render EditExpensePage correctly', () => {
  wrapper = shallow(<EditExpensePage {...{expense, editExpense, deleteExpense, history}}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should redirect when no expense given', () => {
  wrapper = shallow(<EditExpensePage {...{editExpense, deleteExpense, history}}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should handle editExpense', () => {
  wrapper = shallow(<EditExpensePage {...{expense, editExpense, deleteExpense, history}}/>);
  wrapper.find('ExpenseForm').prop('onSubmit')(expense);
  expect(history.push).toHaveBeenLastCalledWith(paths.dashboard);
  expect(editExpense).toHaveBeenLastCalledWith(expense);
});

test('should handle deleteExpense', () => {
  wrapper = shallow(<EditExpensePage {...{expense, editExpense, deleteExpense, history}}/>);
  wrapper.find('button').at(0).simulate('click');
  expect(history.push).toHaveBeenLastCalledWith(paths.dashboard);
  expect(deleteExpense).toHaveBeenLastCalledWith(expense);
});
