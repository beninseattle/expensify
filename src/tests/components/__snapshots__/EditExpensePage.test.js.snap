// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should redirect when no expense given 1`] = `
<Redirect
  push={false}
  to="/dashboard"
/>
`;

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Expense {
        "amount": 1000,
        "createdAt": 1483228800000,
        "description": "abc",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    delete
  </button>
</div>
`;
